Pico HTTP parser support

kellabyte opened the pull request.

nmdguerreiro commented:
Sorry I haven't been very active lately. I will have a look as soon as I can.

aaylward commented on src/haywire/http_connection.h:
```diff
@@ -1,21 +1,24 @@
+//    enum {OPEN, CLOSING, CLOSED} state;
+//    hw_request_buffer* buffer;
+//    unsigned int prevbuflen;
+//} http_connection;
```
howdy -- first time caller here. Any reason not to remove this stuff instead of commenting it out?

aaylward commented on src/haywire/http_request.h:
```diff
@@ -3,9 +3,12 @@
 #pragma once
 #include "uv.h"
 #include "http_parser.h"
+#include "http_connection.h"
```
and then can this include go away?

nmdguerreiro commented on src/haywire/http_server.c:
```diff
@@ -378,6 +385,132 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
+        //fwrite(http_request_buffer_get_buffer(connection->buffer),
+        //       http_request_buffer_get_used(connection->buffer), 1, stdout);
+
+        while (connection->prevbuflen < http_request_buffer_get_used(connection->buffer))
```
@kellabyte I don't get this while loop. Is this so you loop through multiple requests potentially read together? shouldn't it be `while(total_parsed < nread)` , where `total_parsed` is incremented after the call to `phr_parse_request`?

nmdguerreiro commented on src/haywire/http_server.c:
```diff
@@ -378,6 +385,132 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
+            size_t num_headers;
+            num_headers = sizeof(headers) / sizeof(headers[0]);
+
+            parsed = phr_parse_request(&http_request_buffer_get_buffer(connection->buffer)[connection->prevbuflen],
```
If `total_parsed` is incremented after each call to `phr_parse_request` then you could do `buf + total_parsed`. It's simpler to use the pointer passed in as a parameter (i.e. `buf`), as that is guaranteed to be pointing to be beginning of the piece of memory that needs to be parsed.

nmdguerreiro commented on src/haywire/http_server.c:
```diff
@@ -378,6 +385,132 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
+            num_headers = sizeof(headers) / sizeof(headers[0]);
+
+            parsed = phr_parse_request(&http_request_buffer_get_buffer(connection->buffer)[connection->prevbuflen],
+                                       http_request_buffer_get_used(connection->buffer),
```
And this would be `nread - total_parsed`

nmdguerreiro replied:
I don't remember that comment, but I think there are a few issues with the current approach.

1 Looking at signature of `phr_parse_request` into account, `last_len` should tell the parser where
```C
int phr_parse_request(const char *buf_start, size_t len, const char **method, size_t *method_len, const char **path,
                      size_t *path_len, int *minor_version, struct phr_header *headers, size_t *num_headers, size_t last_len)
```
`last_len` is hard-coded to 0. That parameter is used by pico parser to determine if a request has been fully parsed or not. Based on the example at https://github.com/h2o/picohttpparser, it seems like it's supposed to be how far along are we in the buffer after the last round of parsing. Sounds to me like it should really be `http_request_buffer_get_buffer(connection->buffer)->used - nread`, i.e., the bit on the buffer that is yet to be parsed.

2) I don't get the loop in its current form.

`while (connection->prevbuflen < http_request_buffer_get_used(connection->buffer))`

`phr_parse_request` returns a positive number, when it's reached the end of a request. There may be multiple requests in the buffer we're getting, so I think we should loop until all calls from `phr_parser_request` have consumer all `nread` bytes. Example:

```C
int total_parsed = 0;
int parsed;

while (total_parsed < nread) {
  parsed = phr_parser_request(...., total_parsed);
  
  if (parsed > 0) {
     total_parsed += parsed;   
     http_request_buffer_mark(connection->buffer); // so we can sweep the contents of the buffer the next time we get a new chunk of data from libuv
  }
}

http_request_buffer_sweep(connection->buffer); // we've read all there was to read from this buffer, so we can sweep any buffer chunks that are guaranteed not to be in use. 
```

This is as described in:

```c
void http_request_buffer_mark(hw_request_buffer* buf) {
    http_request_buffer* buffer = (http_request_buffer*) buf;
    /* unfortunately, the parser doesn't tell us where the request ends exactly,
     * so the only thing we can be sure is that it ends in the current buffer chunk, so anything before it can
     * effectively be swept, so we're placing the mark at that point now. */
    buffer->mark = buffer->used_before;
}
```

The following example describe how things should be happening for them to work.

Imagine a buffer size of 64KB and two requests of 32KB received 24KB at the time:

1) 64KB are allocated

2) 24KB of request 1 are read into the `hw_request_buffer`. Let's call that chunk 1. There are still 40KB free.

3) The parser parses those 24KB without completing the request. We can't sweep anything from the `hw_request_buffer` buffer as otherwise we'd lose the first 24KB of the first request, so `sweep` is actually a no-op in this case, because a mark hasn't been set yet (i.e. we haven't gotten to the end of a single request so far).

4) We get another 24KB from libuv, on the same physical buffer (hw_request_buffer->buffer), but 24KB from its start. Let's call that chunk 2. 8KB belong to the request 1 and 16KB to request 2.

5) The parser parses 8KB and returns. We place a mark because we now have request 1 in full. This will effectively mark anything before the beginning of chunk #2 as content that can be swept.

6) We still have 16KB to process for the request 2, before processing the next chunk from libuv. We parser parses 16KB, but returns 0 as it hasn't yet finished.

7) We get to the end of the chunk #2 and call sweep. This time, there's a mark, right at the beginning of chunk #2. This will effectively move chunk #2 to the beginning of the physical buffer and overwrite chunk #1.

8) We get another 16KB this time (assume no more requests left), which I'll refer to as chunk #3. The parser runs and returns 16KB. We mark again, therefore saying that anything before the beginning of chunk #3 can be swept now (those first 8KB of the request 1 and 16KB of the request 2).

9) We get to the end of chunk 3, so chunk 2 gets swept away and chunk 3 takes its place in the physical buffer.

I've tried to represent the same sequence in a set of diagrams (sorry, my writing is terrible):

![img_3736 2](https://user-images.githubusercontent.com/2456228/27507886-7aa95668-58d1-11e7-8a9c-289b2c4ba22f.JPG)

nmdguerreiro replied:
Another thing worth mentioning is that whenever the URL, headers or body are discovered by Pico, you'll have to pin the underlying memory region like in https://github.com/haywire/haywire/blob/master/src/haywire/http_request.c#L173, so that if the request buffer is reallocated to a different memory region you can still recover those bits based on their relative position of the buffer, in the new buffer.

nmdguerreiro commented:
@kellabyte I think the difficulty you're facing is that `phr_parse_request` expects you to have a pointer to the beginning of the request whenever the function is called. Reading the source code for `phr_parse_request`, it seems like calling the function as request fragments become available causes a re-parsing of the entire buffer, until such a time when the entire request has been read off the socket.

In our case, things are a bit trickier. When the read callback is called, we get a `char *buf` that points to the beginning of the buffer that libuv used to read the data from the socket. The very first time the read callback is executed that coincides with the beginning of the request, but if for some reason only a fragment of the request is read, you won't get access to the beginning of the request anymore in subsequent calls, and you need that pointer to call `phr_parse_request`.

I think what you need to do is to keep track of the position in the buffer where the last request ended. Therefore, you can always get the beginning of the current request by skipping the number of bytes still taken by the previous request in the buffer. You'd need to make sure that when the buffer is swept, that that pointer is adjusted accordingly though.

On the topic of marking/sweeping, since the other HTTP parser we have doesn't tell us exactly how many bytes were read to reach the end of the buffer, I had to make `mark` use the end of the previous buffer chunk read that contains no unprocessed requests. That was so I could make sure that any requests in-flight split across chunks wouldn't be swept. In this case, since you know exactly where a request ends, you could make mark work differently.

### Example (assuming each request is 100 bytes for simplicity):

#### HTTP parser mark/sweep
* 50 bytes are read
* 20 bytes are read
** We haven't reached the end of a request, so mark isn't called.
* 50 bytes are read 
** We reach the end of the first request with another 30 bytes belonging to the second request. Since we don't know where the first request ended (though we could keep a counter actually), `mark` puts a mark at the end of the previous chunk whenever the first request is fully parsed. In this case, that is 70 bytes. Then when we sweep, we'll get rid of 70 bytes, though 30 of them are useless. Those 30 bytes will only be swept when the second request is parsed successfully.

#### Pico mark/sweep 
* 50 bytes are read
* 20 bytes are read
** We haven't reached the end of a request, so mark isn't called.
* 50 bytes are read 
** After reading 100 bytes from the beginning of the buffer, pico parser returns 100 as is parses the first request successfully. At this stage, we could call the request callback and then set a mark at 100 bytes and instantly call sweep. That should move the remaining 30 bytes back to the beginning of the buffer, causing subsequent alloc calls to start re-using the buffer beyond the 30 bytes that the buffer is currently keeping.

kellabyte commented:
Isn't the code already calling `mark` and `sweep` in the way you're describing or am I miss understanding the sequence you're describing?

kellabyte commented:
The current code has the following benchmark results.  It seems to stall. For some reason `PREVBUFLEN` looks to be reset.

```
../bin/wrk/wrk --script ./pipelined_get.lua --latency -d 1s -t 1 -c 1 http://127.0.0.1:8000 -- 64
Running 1s test @ http://127.0.0.1:8000
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.41ms    0.00us   5.41ms  100.00%
    Req/Sec   640.00      0.00   640.00    100.00%
  Latency Distribution
     50%    5.41ms
     75%    5.41ms
     90%    5.41ms
     99%    5.41ms
  66 requests in 1.03s, 8.38KB read
Requests/sec:     63.87
Transfer/sec:      8.11KB

Added route /ping
Added route /
Address: 0.0.0.0
Port: 8000
Threads: 0
Parser: pico
TCP No Delay: off
Listen backlog: 128
Maximum request size: 1048576
Listening...
START	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 0
LOOP	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 0
SWEEP	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 78 OLD: 78 NEW: 78 PREVBUFLEN: 78
START	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 0
LOOP	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 0
SWEEP	NREAD: 78 BUFUSED: 78 PREVBUFLEN: 78 OLD: 78 NEW: 78 PREVBUFLEN: 78
START	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 0
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 0
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 40 OLD: 2560 NEW: 2560 PREVBUFLEN: 40
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 40
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 80 OLD: 2560 NEW: 2560 PREVBUFLEN: 80
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 80
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 120 OLD: 2560 NEW: 2560 PREVBUFLEN: 120
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 120
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 160 OLD: 2560 NEW: 2560 PREVBUFLEN: 160
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 160
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 200 OLD: 2560 NEW: 2560 PREVBUFLEN: 200
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 200
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 240 OLD: 2560 NEW: 2560 PREVBUFLEN: 240
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 240
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 280 OLD: 2560 NEW: 2560 PREVBUFLEN: 280
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 280
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 320 OLD: 2560 NEW: 2560 PREVBUFLEN: 320
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 320
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 360 OLD: 2560 NEW: 2560 PREVBUFLEN: 360
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 360
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 400 OLD: 2560 NEW: 2560 PREVBUFLEN: 400
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 400
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 440 OLD: 2560 NEW: 2560 PREVBUFLEN: 440
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 440
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 480 OLD: 2560 NEW: 2560 PREVBUFLEN: 480
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 480
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 520 OLD: 2560 NEW: 2560 PREVBUFLEN: 520
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 520
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 560 OLD: 2560 NEW: 2560 PREVBUFLEN: 560
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 560
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 600 OLD: 2560 NEW: 2560 PREVBUFLEN: 600
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 600
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 640 OLD: 2560 NEW: 2560 PREVBUFLEN: 640
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 640
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 680 OLD: 2560 NEW: 2560 PREVBUFLEN: 680
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 680
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 720 OLD: 2560 NEW: 2560 PREVBUFLEN: 720
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 720
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 760 OLD: 2560 NEW: 2560 PREVBUFLEN: 760
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 760
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 800 OLD: 2560 NEW: 2560 PREVBUFLEN: 800
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 800
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 840 OLD: 2560 NEW: 2560 PREVBUFLEN: 840
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 840
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 880 OLD: 2560 NEW: 2560 PREVBUFLEN: 880
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 880
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 920 OLD: 2560 NEW: 2560 PREVBUFLEN: 920
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 920
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 960 OLD: 2560 NEW: 2560 PREVBUFLEN: 960
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 960
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1000 OLD: 2560 NEW: 2560 PREVBUFLEN: 1000
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1000
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1040 OLD: 2560 NEW: 2560 PREVBUFLEN: 1040
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1040
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1080 OLD: 2560 NEW: 2560 PREVBUFLEN: 1080
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1080
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1120 OLD: 2560 NEW: 2560 PREVBUFLEN: 1120
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1120
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1160 OLD: 2560 NEW: 2560 PREVBUFLEN: 1160
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1160
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1200 OLD: 2560 NEW: 2560 PREVBUFLEN: 1200
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1200
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1240 OLD: 2560 NEW: 2560 PREVBUFLEN: 1240
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1240
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1280 OLD: 2560 NEW: 2560 PREVBUFLEN: 1280
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1280
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1320 OLD: 2560 NEW: 2560 PREVBUFLEN: 1320
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1320
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1360 OLD: 2560 NEW: 2560 PREVBUFLEN: 1360
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1360
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1400 OLD: 2560 NEW: 2560 PREVBUFLEN: 1400
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1400
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1440 OLD: 2560 NEW: 2560 PREVBUFLEN: 1440
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1440
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1480 OLD: 2560 NEW: 2560 PREVBUFLEN: 1480
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1480
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1520 OLD: 2560 NEW: 2560 PREVBUFLEN: 1520
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1520
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1560 OLD: 2560 NEW: 2560 PREVBUFLEN: 1560
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1560
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1600 OLD: 2560 NEW: 2560 PREVBUFLEN: 1600
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1600
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1640 OLD: 2560 NEW: 2560 PREVBUFLEN: 1640
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1640
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1680 OLD: 2560 NEW: 2560 PREVBUFLEN: 1680
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1680
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1720 OLD: 2560 NEW: 2560 PREVBUFLEN: 1720
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1720
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1760 OLD: 2560 NEW: 2560 PREVBUFLEN: 1760
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1760
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1800 OLD: 2560 NEW: 2560 PREVBUFLEN: 1800
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1800
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1840 OLD: 2560 NEW: 2560 PREVBUFLEN: 1840
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1840
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1880 OLD: 2560 NEW: 2560 PREVBUFLEN: 1880
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1880
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1920 OLD: 2560 NEW: 2560 PREVBUFLEN: 1920
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1920
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1960 OLD: 2560 NEW: 2560 PREVBUFLEN: 1960
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 1960
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2000 OLD: 2560 NEW: 2560 PREVBUFLEN: 2000
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2000
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2040 OLD: 2560 NEW: 2560 PREVBUFLEN: 2040
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2040
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2080 OLD: 2560 NEW: 2560 PREVBUFLEN: 2080
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2080
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2120 OLD: 2560 NEW: 2560 PREVBUFLEN: 2120
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2120
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2160 OLD: 2560 NEW: 2560 PREVBUFLEN: 2160
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2160
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2200 OLD: 2560 NEW: 2560 PREVBUFLEN: 2200
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2200
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2240 OLD: 2560 NEW: 2560 PREVBUFLEN: 2240
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2240
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2280 OLD: 2560 NEW: 2560 PREVBUFLEN: 2280
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2280
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2320 OLD: 2560 NEW: 2560 PREVBUFLEN: 2320
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2320
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2360 OLD: 2560 NEW: 2560 PREVBUFLEN: 2360
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2360
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2400 OLD: 2560 NEW: 2560 PREVBUFLEN: 2400
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2400
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2440 OLD: 2560 NEW: 2560 PREVBUFLEN: 2440
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2440
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2480 OLD: 2560 NEW: 2560 PREVBUFLEN: 2480
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2480
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2520 OLD: 2560 NEW: 2560 PREVBUFLEN: 2520
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2520
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2560 OLD: 2560 NEW: 2560 PREVBUFLEN: 2560
START	NREAD: 2560 BUFUSED: 5120 PREVBUFLEN: 2560
LOOP	NREAD: 2560 BUFUSED: 5120 PREVBUFLEN: 2560
SWEEP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 2600 OLD: 5120 NEW: 2560 PREVBUFLEN: 40
LOOP	NREAD: 2560 BUFUSED: 2560 PREVBUFLEN: 40
SWEEP	NREAD: 2560 BUFUSED: 0 PREVBUFLEN: 80 OLD: 2560 NEW: 0 PREVBUFLEN: 0
```

kellabyte commented:
@nmdguerreiro I would appreciate a conversation about this on IRC sometime if you have any time to spare that would be great :) I think you still have a better understanding of how the buffer manager and libuv works than I do.

nmdguerreiro commented on src/haywire/http_request_buffers.c:
```diff
@@ -237,3 +237,13 @@ void http_request_buffer_destroy(hw_request_buffer* buf) {
     free(buffer);
 }
+
+void* http_request_buffer_get_buffer(hw_request_buffer* buf) {
```
Maybe turn this into a macro or inline it, so this doesn't result in a function call.
